File: designated_customer_flutter/lib/features/home/widgets/step_counter_card.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

/// 메인 화면 만보기 카드 - 가로 레이아웃 (TODAY 버튼 - 원형 - SESSION 버튼/리셋)
class StepCounterCard extends StatefulWidget {
  const StepCounterCard({super.key});

  @override
  State<StepCounterCard> createState() => _StepCounterCardState();
}

class _StepCounterCardState extends State<StepCounterCard> {
  bool _isSessionMode = false;
  final int _todaySteps = 5234;
  final int _sessionSteps = 1250;
  final int _goal = 10000;

  @override
  Widget build(BuildContext context) {
    final todayProgress = (_todaySteps / _goal).clamp(0.0, 1.0);
    final sessionGoal = _goal / 2;
    final sessionProgress = (_sessionSteps / sessionGoal).clamp(0.0, 1.0);
    final displaySteps = _isSessionMode ? _sessionSteps : _todaySteps;
    final displayProgress = _isSessionMode ? sessionProgress : todayProgress;
    final displayLabel = _isSessionMode ? 'SESSION' : 'TODAY';

    return Card(
      elevation: 2,
      margin: EdgeInsets.zero,
      shape: const RoundedRectangleBorder(borderRadius: BorderRadius.zero),
      color: const Color(0xFF2C2C2C),
      child: SizedBox(
        height: 200,
        child: Stack(
          children: [
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 24),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Expanded(
                    flex: 1,
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Text('TODAY', style: TextStyle(fontSize: 11, fontWeight: FontWeight.bold, color: !_isSessionMode ? const Color(0xFFFFAB00) : Colors.grey)),
                        const SizedBox(height: 4),
                        _buildCircleButton(steps: _todaySteps, isActive: !_isSessionMode, color: const Color(0xFFFFAB00), onPressed: () => setState(() => _isSessionMode = false)),
                      ],
                    ),
                  ),
                  Expanded(flex: 2, child: Center(child: _CircularStepCounter(steps: displaySteps, progress: displayProgress, label: displayLabel))),
                  Expanded(
                    flex: 1,
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Text('SESSION', style: TextStyle(fontSize: 11, fontWeight: FontWeight.bold, color: _isSessionMode ? const Color(0xFF00BCD4) : Colors.grey)),
                        const SizedBox(height: 4),
                        _buildCircleButton(steps: _sessionSteps, isActive: false, color: const Color(0xFF00BCD4), onPressed: () => setState(() => _isSessionMode = true)),
                      ],
                    ),
                  ),
                ],
              ),
            ),
            Positioned(
              top: 4, right: 4,
              child: IconButton(onPressed: () {}, icon: const Icon(Icons.settings, size: 18, color: Color(0x80FFFFFF)), padding: EdgeInsets.zero, constraints: const BoxConstraints(minWidth: 32, minHeight: 32)),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildCircleButton({required int steps, required bool isActive, required Color color, required VoidCallback onPressed}) {
    return SizedBox(
      width: 80, height: 80,
      child: OutlinedButton(
        onPressed: onPressed,
        style: OutlinedButton.styleFrom(shape: const CircleBorder(), side: BorderSide(width: 3, color: isActive ? color : Colors.grey), backgroundColor: isActive ? color.withOpacity(0.2) : Colors.transparent, padding: EdgeInsets.zero),
        child: Text(_formatSteps(steps), style: TextStyle(fontSize: 14, fontWeight: FontWeight.bold, color: isActive ? color : Colors.grey)),
      ),
    );
  }

  String _formatSteps(int steps) {
    if (steps >= 1000) return NumberFormat('#,###').format(steps);
    return steps.toString();
  }
}

class _CircularStepCounter extends StatelessWidget {
  final int steps;
  final double progress;
  final String label;

  const _CircularStepCounter({required this.steps, required this.progress, required this.label});

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 140, height: 140,
      child: Stack(
        alignment: Alignment.center,
        children: [
          CustomPaint(size: const Size(140, 140), painter: _CircularProgressPainter(progress: progress)),
          Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Text('👟', style: TextStyle(fontSize: 24)),
              const SizedBox(height: 4),
              Text(_formatSteps(steps), style: const TextStyle(fontSize: 26, fontWeight: FontWeight.bold, color: Color(0xFF00BCD4), letterSpacing: 0)),
              const SizedBox(height: 2),
              Text(label, style: const TextStyle(fontSize: 10, fontWeight: FontWeight.w500, color: Color(0x99FFFFFF), letterSpacing: 0.5)),
            ],
          ),
        ],
      ),
    );
  }

  String _formatSteps(int steps) {
    if (steps >= 1000) return NumberFormat('#,###').format(steps);
    return steps.toString();
  }
}

class _CircularProgressPainter extends CustomPainter {
  final double progress;
  const _CircularProgressPainter({required this.progress});

  @override
  void paint(Canvas canvas, Size size) {
    const orangeColor = Color(0xFFFFAB00);
    const lightGray = Color(0xFFE0E0E0);
    const strokeWidth = 10.0;
    final center = Offset(size.width / 2, size.height / 2);
    final radius = (size.width - strokeWidth) / 2;

    canvas.drawCircle(center, radius, Paint()..color = lightGray..style = PaintingStyle.stroke..strokeWidth = strokeWidth..strokeCap = StrokeCap.round);

    if (progress > 0) {
      final sweepAngle = 2 * math.pi * progress;
      canvas.drawArc(Rect.fromCircle(center: center, radius: radius), -math.pi / 2, sweepAngle, false, Paint()..color = orangeColor..style = PaintingStyle.stroke..strokeWidth = strokeWidth..strokeCap = StrokeCap.round);
    }
  }

  @override
  bool shouldRepaint(_CircularProgressPainter oldDelegate) => oldDelegate.progress != progress;
}

File: customer_app/app/src/main/java/com/designated/customer/ui/components/StepCounterCard.kt
package com.designated.customer.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.draw.blur
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.designated.customer.R
import com.designated.customer.data.model.DailyStepData

/**
 * 메인 화면 만보기 카드 - 가로 레이아웃 (TODAY 버튼 - 원형 - 리셋 버튼)
 */
@Composable
fun StepCounterCard(
    stepData: DailyStepData?,
    sessionSteps: Int = 0,
    isSessionActive: Boolean = false,
    onSettingsClick: () -> Unit,
    onStartSession: () -> Unit = {},
    onResetSession: () -> Unit = {},
    modifier: Modifier = Modifier
) {
    // 모드 상태 (false: TODAY, true: SESSION)
    var isSessionMode by remember { mutableStateOf(false) }

    val todaySteps = stepData?.steps ?: 0
    val goal = stepData?.goal ?: 10000
    val todayProgress = if (goal > 0) (todaySteps.toFloat() / goal.toFloat()).coerceIn(0f, 1f) else 0f

    // 세션 진행률은 목표의 절반 기준
    val sessionGoal = goal / 2
    val sessionProgress = if (sessionGoal > 0) (sessionSteps.toFloat() / sessionGoal.toFloat()).coerceIn(0f, 1f) else 0f

    // 현재 표시할 걸음수와 진행률
    val displaySteps = if (isSessionMode) sessionSteps else todaySteps
    val displayProgress = if (isSessionMode) sessionProgress else todayProgress
    val displayLabel = if (isSessionMode) "SESSION" else "TODAY"

    Card(
        modifier = modifier
            .fillMaxWidth()
            .height(200.dp),
        shape = RoundedCornerShape(0.dp),
        colors = CardDefaults.cardColors(
            containerColor = Color.Transparent
        ),
        elevation = CardDefaults.cardElevation(defaultElevation = 0.dp)
    ) {
        Box(
            modifier = Modifier.fillMaxSize()
        ) {
            // 배경 이미지
            Image(
                painter = painterResource(id = R.drawable.step_counter_background),
                contentDescription = "만보기 배경",
                modifier = Modifier
                    .fillMaxSize()
                    .blur(3.dp),
                contentScale = ContentScale.Crop
            )

            // 메인 가로 레이아웃
            Row(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(horizontal = 16.dp, vertical = 24.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                // 좌측: TODAY 원형 버튼과 텍스트
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.Center,
                    modifier = Modifier.weight(1f)
                ) {
                    Text(
                        text = "TODAY",
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Bold,
                        color = if (!isSessionMode) Color(0xFFFFAB00) else Color.Gray
                    )
                    Spacer(modifier = Modifier.height(4.dp))
                    OutlinedButton(
                        onClick = {
                            isSessionMode = false
                        },
                        colors = ButtonDefaults.outlinedButtonColors(
                            containerColor = if (!isSessionMode) Color(0xFFFFAB00).copy(alpha = 0.2f) else Color.Transparent,
                            contentColor = if (!isSessionMode) Color(0xFFFFAB00) else Color.Gray
                        ),
                        border = BorderStroke(
                            width = 3.dp,
                            color = if (!isSessionMode) Color(0xFFFFAB00) else Color.LightGray
                        ),
                        modifier = Modifier.size(80.dp),
                        shape = CircleShape,
                        contentPadding = PaddingValues(0.dp)
                    ) {
                        Text(
                            text = formatSteps(todaySteps),
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }

                // 중앙: 원형 카운터 (큰 원)
                Box(
                    modifier = Modifier.weight(1.5f),
                    contentAlignment = Alignment.Center
                ) {
                    CircularStepCounter(
                        steps = displaySteps,
                        progress = displayProgress,
                        label = displayLabel
                    )
                }

                // 우측: SESSION 원형 버튼/리셋과 텍스트
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.Center,
                    modifier = Modifier.weight(1f)
                ) {
                    Text(
                        text = "SESSION",
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Bold,
                        color = if (isSessionMode) Color(0xFF00BCD4) else Color.Gray
                    )
                    Spacer(modifier = Modifier.height(4.dp))
                    if (!isSessionActive || !isSessionMode) {
                        // SESSION 시작 버튼
                        OutlinedButton(
                            onClick = {
                                onStartSession()
                                isSessionMode = true
                            },
                            colors = ButtonDefaults.outlinedButtonColors(
                                containerColor = Color.Transparent,
                                contentColor = Color(0xFF00BCD4)
                            ),
                            border = BorderStroke(
                                width = 3.dp,
                                color = Color(0xFF00BCD4)
                            ),
                            modifier = Modifier.size(80.dp),
                            shape = CircleShape,
                            contentPadding = PaddingValues(0.dp)
                        ) {
                            Text(
                                text = formatSteps(sessionSteps),
                                fontSize = 14.sp,
                                fontWeight = FontWeight.Bold
                            )
                        }
                    } else {
                        // 리셋 버튼
                        OutlinedButton(
                            onClick = onResetSession,
                            colors = ButtonDefaults.outlinedButtonColors(
                                containerColor = Color(0xFF00BCD4).copy(alpha = 0.2f),
                                contentColor = Color(0xFF00BCD4)
                            ),
                            border = BorderStroke(
                                width = 3.dp,
                                color = Color(0xFF00BCD4)
                            ),
                            modifier = Modifier.size(80.dp),
                            shape = CircleShape,
                            contentPadding = PaddingValues(0.dp)
                        ) {
                            Column(
                                horizontalAlignment = Alignment.CenterHorizontally,
                                verticalArrangement = Arrangement.Center
                            ) {
                                Icon(
                                    imageVector = Icons.Default.Refresh,
                                    contentDescription = "리셋",
                                    modifier = Modifier.size(28.dp)
                                )
                                Spacer(modifier = Modifier.height(2.dp))
                                Text(
                                    text = formatSteps(sessionSteps),
                                    fontSize = 12.sp,
                                    fontWeight = FontWeight.Bold
                                )
                            }
                        }
                    }
                }
            }

            // 우측 상단 설정 아이콘
            IconButton(
                onClick = onSettingsClick,
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .padding(4.dp)
                    .size(32.dp)
            ) {
                Icon(
                    imageVector = Icons.Default.Settings,
                    contentDescription = "만보기 설정",
                    tint = Color.White.copy(alpha = 0.7f),
                    modifier = Modifier.size(18.dp)
                )
            }
        }
    }
}

/**
 * 원형 걸음 수 표시 컴포넌트
 */
@Composable
private fun CircularStepCounter(
    steps: Int,
    progress: Float,
    label: String,
    modifier: Modifier = Modifier
) {
    val orangeColor = Color(0xFFFFAB00)
    val lightGray = Color(0xFFE0E0E0)

    Box(
        modifier = modifier.size(140.dp),
        contentAlignment = Alignment.Center
    ) {
        // Canvas로 원형 프로그레스 그리기
        Canvas(
            modifier = Modifier.size(140.dp)
        ) {
            val canvasSize = size.minDimension
            val strokeWidth = 10.dp.toPx()
            val arcSize = Size(canvasSize - strokeWidth, canvasSize - strokeWidth)
            val topLeft = Offset(strokeWidth / 2, strokeWidth / 2)

            // 배경 원 (밝은 회색)
            drawArc(
                color = lightGray,
                startAngle = -90f,
                sweepAngle = 360f,
                useCenter = false,
                topLeft = topLeft,
                size = arcSize,
                style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
            )

            // 프로그레스 원 (오렌지)
            if (progress > 0f) {
                drawArc(
                    color = orangeColor,
                    startAngle = -90f,
                    sweepAngle = 360f * progress,
                    useCenter = false,
                    topLeft = topLeft,
                    size = arcSize,
                    style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
                )
            }
        }

        // 중앙 콘텐츠
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            // 걸음 수
            Text(
                text = formatSteps(steps),
                fontSize = 26.sp,
                fontWeight = FontWeight.Bold,
                color = Color(0xFF00BCD4),
                letterSpacing = 0.sp
            )

            Spacer(modifier = Modifier.height(2.dp))

            // 라벨
            Text(
                text = label,
                fontSize = 10.sp,
                fontWeight = FontWeight.Medium,
                color = Color.White.copy(alpha = 0.8f),
                letterSpacing = 0.5.sp
            )
        }
    }
}

/**
 * 걸음 수 포맷팅 (천 단위 구분)
 */
private fun formatSteps(steps: Int): String {
    return when {
        steps >= 10000 -> String.format("%,d", steps)
        steps >= 1000 -> String.format("%,d", steps)
        else -> steps.toString()
    }
}
